### ISSUE TITLE
failed autojoin when joining a complex condition

### ISSUE THREAD

**Issue by alissapajer** [quasar-analytics/quasar#1867]
In the integration test "join with complex condition" (`joins/joinWithComplexCondition.test`) we fail to `autojoin` in QScript.

This behavior is exhibited after the merge of #1865.

```
               "select l.sha as child, l.author.login as c_auth, r.sha as parent, r.author.login as p_auth
                from `../slamengine_commits` as l join `../slamengine_commits_dup` as r
                on r.sha = l.parents[0].sha and l.author.login = r.author.login
                where l.author.login || \",\" || r.author.login = \"selloutsellout\""
```

**Comment by alissapajer** [comment]
Simplified examples:
```
select r.x, l.y from slamengine_commits as l join slamengine_commits_dup as r on r.x=l.a.b
```
and
```
select r.x, l.y from slamengine_commits as l join slamengine_commits_dup as r on r.x=l.a[0]
```
When the join condition projects more than once into `l` (e.g. `l.a.b` or `l.a[0]`) we fail to autojoin.

Digging in a bit, the problem happens before we even try to construct the actual `ThetaJoin`. We first construct a `ThetaJoin` that just represents `r.x = l.a.b`, which we expect to do. But the construction of this has yet another `ThetaJoin` as one of _its_ branches, which is what's incorrect.

**Comment by alissapajer** [comment]
What QScript is failing to support is illustrated really nicely when you look at the LogicalPlan. Consider the join condition `select * from zips join smallZips on zips.ack = smallZips.foo`. Then in the LP we have
```
      ├─ Invoke(InnerJoin)
      │  ├─ Free('__tmp0)
      │  ├─ Free('__tmp1)
      │  ╰─ Invoke(Eq)
      │     ├─ ObjectProject('__tmp0{"x"})
      │     ╰─ ObjectProject('__tmp1{"foo"})
```
In this first case, we join the left and right branches of the `Eq` with `'__tmp0` and `'__tmp1`, respectively, and all merges cleanly.

Now consider `select * from zips join smallZips on zips.ack = smallZips.foo.bar` whose pertinent LP subset looks like
```
      ├─ Invoke(InnerJoin)
      │  ├─ Free('__tmp0)
      │  ├─ Free('__tmp1)
      │  ╰─ Let('__check2)
      │     ├─ ObjectProject('__tmp1{"foo"})
      │     ╰─ Typecheck(Obj(Map[], Some(Top)))
      │        ├─ Free('__check2)
      │        ├─ Invoke(Eq)
      │        │  ├─ ObjectProject('__tmp0{"x"})
      │        │  ╰─ ObjectProject('__check2{"bar"}) 
      │        ╰─ Constant(NA)
```
In this second case, we still need to join the left and right branches of the condition with `'__tmp0` and `'__tmp1`, but now, written in shorthand, the condition is really
```
Typecheck(
  Obj,
  tmp1.foo,
  Eq(tmp0.x, tmp1.foo.bar),
  Undefined)
```
We do not "look through" the `Typecheck` to look at the branches of the `Eq` and instead take the `Typecheck` as the whole join condition, which in QScript will contain an extra `ThetaJoin` (from the `autojoin3` that attempts to join the 3 sides of the `Typecheck`).

**Comment by sellout** [comment]
Oh! We fixed[†] this on @drostron’s key join branch. It could perhaps be extracted?

[†]: I’m sure it needs some massaging to not break anything.

**Comment by alissapajer** [comment]
@sellout Oh! What's the status of that branch? Is it pushed? Did you come to same conclusion about the problem as I did above? 

What about #1866 - did you folks look into that one at all? (Could be the same problem, I haven't looked at all yet.)

**Comment by alissapajer** [comment]
@sellout I'd be happy to extract it if possible. I'd like to be sure these QScript bugs are out of the way before I dive into performance work.

**Comment by sellout** [comment]
Yeah, it looks like that might be the same. The change in @drostron’s branch is a real fix to the problem. There is also a crappy fix that we used in the old Mongo planner that is called something like `elideTypechecksFromJoinCondition`, which just throws away the Typechecks that get in the way.

**Comment by alissapajer** [comment]
Cool. @drostron what's your branch? Is this bug fix a separate commit or mixed in with other stuff? If I could get that work, that would be awesome. I'd be happy to extract it myself. Or if you're close to PRing, then this ticket can wait until then.

**Comment by alissapajer** [comment]
(Note: test `joinWithMultipleConditions.test` and `joinWithReversedCondition.test` also have this same problem. When this is fixed, it should be verified that those tests also work.)

**Comment by drostron** [comment]
@alissapajer Please excuse the exploratory condition of the branch, but here's the [relevant commit](https://github.com/drostron/quasar/blob/9561b42d316553ccff7816d4747d41ff7608c68e/frontend/src/main/scala/quasar/frontend/logicalplan/Optimizer.scala). Files of interest are `quasar/package.scala`, `LogicalPlanR.scala`, `Optimizer.scala`, and `quasar/sql/helpers.scala`.

**Comment by drostron** [comment]
Updated ult with branch and relevant commit.

**Comment by alissapajer** [comment]
@drostron thanks! I'll let you know what I end up doing so you can merge any work I do back into your branch hopefully easily.

**Comment by sellout** [comment]
Ok, so a bit of explanation … the most important bit is [the Typecheck cases here](https://github.com/drostron/quasar/commit/9561b42d316553ccff7816d4747d41ff7608c68e#diff-67ad2a826d4dc53505ed25ff0beedf3aR254). What got fixed in this code is that Typechecks _used to_ trigger a condition being pulled into a post-Filter (because we never had ThetaJoins), we needed to keep the equality as a join condition for CB joins to work, so now Typechecks get applied to one branch or the other as a pre-filter, leaving only the actual equality condition in the join.

As an aside, the part of this code that moved non-equi conditions to post filters should be removed, since we don’t want to do that unless a connector tells us to convert ThetaJoins to EquiJoins, but it can’t be removed until old Mongo is gone.
